File: TableGames.Games.Doubt/DoubtInitial.cs
using System.Collections.Generic;
using System.Linq;
using TableGames.Domain;

namespace TableGames.Games.Doubt
{
    [GameDescriptor("DoubtInitial", "doubtInitial")]
    public class DoubtInitial : Game
    {
        public List<PlayerCup> PlayerCups { get; set; }

        public DoubtInitial(Table table) : base(table) {
            PlayerCups = new List<PlayerCup>(Table.Players.Select(p => new PlayerCup(p)));

            rollDices(PlayerCups);
            var maxDiceValue = PlayerCups.Select(pc => pc.Dices.Last().Value).Max();
            var maxPlayerCups = PlayerCups.Where(pc => pc.Dices.Last().Value == maxDiceValue).ToList();

            while (maxPlayerCups.Count() > 1) {
                rollDices(maxPlayerCups);
                maxDiceValue = maxPlayerCups.Select(pc => pc.Dices.Last().Value).Max();
                maxPlayerCups = PlayerCups.Where(pc => pc.Dices.Last().Value == maxDiceValue).ToList();
            }

            var winner = maxPlayerCups.First().Player;
            Winners.Add(winner);
            Table.SetNextPlayer(winner);

            IsEnded = true;
        }

        public override object ToClient() {
            return new {
                playerCups = PlayerCups.Select(pc => pc.ToClient()),
                table = new {
                    activePlayerName = Table.ActivePlayer?.Name,
                },
                isEnded = IsEnded,
                winnerNames = Winners.Select(p => p.Name)
            };
        }

        public override object ToStats() {
            return null;
        }

        private void rollDices(List<PlayerCup> playerCups) {
            foreach (var playerCup in playerCups) {
                playerCup.Dices.Add(new Dice(true) { IsExposed = true });
            }
        }
    }
}

File: TableGames.Games.Doubt/doubtInitial.js
define(['knockout', 'jquery', 'tmpl!Doubt/doubtInitial', 'game!Doubt/dice', 'game!Doubt/playerCup'], function(ko, $, doubtInitialTemplateName, Dice, PlayerCup) {

    function DoubtInitial(gameConfig, gameState, table) {
        var self = this;
        var playerPromises = {};

        // initialize player cups
        self.playerCups = gameState.playerCups.map(function(playerCupState) {
            playerPromises[playerCupState.playerName] = $.when();

            var playerCup = new PlayerCup({
                playerName: playerCupState.playerName,
                dices: []
            });

            // ... add dices
            playerCupState.dices.forEach(function(diceState) {
                var dice = new Dice(diceState);
                if (!gameState.inProgress) {
                    playerPromises[playerCupState.playerName] = playerPromises[playerCupState.playerName]
                        .then(function() {
                            playerCup.dices.push(dice);
                            return dice.roll(diceState.value);
                        })
                        .then(delay);
                } else {
                    playerCup.dices.push(dice);
                }
            });

            return playerCup;
        });

        // set active player
        self.isEnded = ko.observable(false);
        playerPromises[gameState.table.activePlayerName].then(function() {
            var winnerCup = ko.utils.arrayFirst(self.playerCups, function(pc) { return pc.playerName === gameState.table.activePlayerName; });
            var winnerLastDice = winnerCup.dices()[winnerCup.dices().length - 1];
            winnerLastDice.isHighlighted(true);

            table.activePlayerName(gameState.table.activePlayerName);
            self.isEnded(gameState.isEnded);
        });
    }
    DoubtInitial.templateName = doubtInitialTemplateName;

    function delay() {
        var dfd = $.Deferred();
        setTimeout(function() { dfd.resolve(); }, 1000);
        return dfd.promise();
    }

    return DoubtInitial;
});
